src/fix/main.c

```diff
@@ -65,535 +75,946 @@ static void print_usage(void)
-			break;
-		case 'i':
-			setid = true;
+	MBC2 = 0x05,
+	MBC2_BATTERY = 0x06,
 
-			if (strlen(optarg) != 4)
-				errx(1, "Game ID %s must be exactly 4 characters",
-				     optarg);
+	MMM01 = 0x0B,
+	MMM01_RAM = 0x0C,
+	MMM01_RAM_BATTERY = 0x0D,
 
-			id = optarg;
-			break;
-		case 'j':
-			nonjapan = true;
-			break;
-		case 'k':
-			setnewlicensee = true;
+	MBC3 = 0x11,
+	MBC3_TIMER_BATTERY = 0x0F,
+	MBC3_TIMER_RAM_BATTERY = 0x10,
+	MBC3_RAM = 0x12,
+	MBC3_RAM_BATTERY = 0x13,
 
-			if (strlen(optarg) != 2)
-				errx(1, "New licensee code %s is not the correct length of 2 characters",
-				     optarg);
+	MBC5 = 0x19,
+	MBC5_RAM = 0x1A,
+	MBC5_RAM_BATTERY = 0x1B,
+	MBC5_RUMBLE = 0x1C,
+	MBC5_RUMBLE_RAM = 0x1D,
+	MBC5_RUMBLE_RAM_BATTERY = 0x1E,
 
-			newlicensee = optarg;
-			break;
-		case 'l':
-			setlicensee = true;
+	MBC6 = 0x20,
 
-			licensee = strtoul(optarg, &ep, 0);
-			if (optarg[0] == '\0' || *ep != '\0')
-				errx(1, "Invalid argument for option 'l'");
+	MBC7_SENSOR_RUMBLE_RAM_BATTERY = 0x22,
 
-			if (licensee < 0 || licensee > 0xFF)
-				errx(1, "Argument for option 'l' must be between 0 and 255");
+	POCKET_CAMERA = 0xFC,
 
-			break;
-		case 'm':
-			setcartridge = true;
+	BANDAI_TAMA5 = 0xFD,
 
-			cartridge = parse_cartridge(optarg, &ep);
-			if (optarg[0] == '\0' || *ep != '\0')
-				errx(1, "Invalid argument for option 'm'");
+	HUC3 = 0xFE,
 
-			if (cartridge < 0 || cartridge > 0xFF)
-				errx(1, "Argument for option 'm' must be between 0 and 255");
+	HUC1_RAM_BATTERY = 0xFF,
 
-			break;
-		case 'n':
-			setversion = true;
+	// Error values
+	MBC_NONE = 0x100, // No MBC specified, do not act on it
+	MBC_BAD = 0x101, // Specified MBC does not exist / syntax error
+	MBC_FEATURES = 0x102, // MBC incompatible with specified features
+	MBC_RANGE = 0x103, // MBC number out of range
+};
 
-			version = strtoul(optarg, &ep, 0);
+/**
+ * @return False on failure
+ */
+static bool readMBCSlice(char const **name, char const *expected)
+{
+	while (*expected) {
+		char c = *(*name)++;
 
-			if (optarg[0] == '\0' || *ep != '\0')
-				errx(1, "Invalid argument for option 'n'");
+		if (c == '\0') // Name too short
+			return false;
 
-			if (version < 0 || version > 0xFF)
-				errx(1, "Argument for option 'n' must be between 0 and 255");
+		if (c >= 'a' && c <= 'z') // Perform the comparison case-insensitive
+			c = c - 'a' + 'A';
+		else if (c == '_') // Treat spaces are underscores
+			c = ' ';
 
+		if (c != *expected++)
+			return false;
+	}
+	return true;
+}
+
+static enum MbcType parseMBC(char const *name)
+{
+	if (name[0] >= '0' && name[0] <= '9') {
+		// Parse number, and return it as-is (unless it's too large)
+		char *endptr;
+		unsigned long mbc = strtoul(name, &endptr, 0);
+
+		if (*endptr)
+			return MBC_BAD;
+		if (mbc > 0xFF)
+			return MBC_RANGE;
+		return mbc;
+
+	} else {
+		// Begin by reading the MBC type:
+		uint16_t mbc;
+		char const *ptr = name;
+
+		// Trim off leading whitespace
+		while (*name == ' ')
```

Comment:
Also check for `'\t'` if the docs claim "whitespace" is ignored.

Reply:
Might as well use `isblank`? It's POSIX 2001 and C99, so it should be supported even by MSVC.

Reply:
`isblank` considers only space and tab; `isspace` also considers `\n\v\f\r`. If you just want to support tab and space, I'd just do `*name == ' ' || *name == '\t'`, since you also do explicit comparison with `'0'`/`'9'` instead of `isdigit`. (I think `isdigit` is only local-sensitive in C++ from `<locale>`; C's is ASCII.)

Reply:
C's `isdigit` and friends are just as locale-sensitive as C++'s. It's just that, in C, the default locale is the C locale, unless you explicitly change it with a call to `setlocale`; the environment is _not allowed_ to default to something else. (I don't know about C++'s semantics.)

Reply:
I see. IMO stick with explicit `' '`/`'\t'`/`'0'`/`'9'` comparisons then.